### PR TITLE
#264 Implement is_nil & is_a Method for Type Checking

### PR DESCRIPTION
**Pull request by Maxwell-Alexius**

### CODE REVIEW

File: vm/class_test.go [hunk 1]
```diff
@@ -453,4 +462,123 @@ func TestRequireFail(t *testing.T) {
...
+		{`[1, 2, 3, 4, 5].is_a(Array)`, true},
+		{`[1, 2, 3, 4, 5].is_a(Object)`, true},
+		{`[1, 2, 3, 4, 5].is_a(Null)`, false},
+		{`[1, 2, 3, 4, 5].is_a(String)`, false},
+		{`true.is_a(Boolean)`, true},
+		{`true.is_a(Object)`, true},
+		{`true.is_a(Array)`, false},
+		{`true.is_a(Integer)`, false},
+		// TODO: Fix Class.is_a Error when argument is a
+		//{`String.is_a(Class)`, true},
```
Comment (Maxwell-Alexius) [inline vm/class_test.go]: In this test case, I try it in REPL and got the result:

```
$ go run goby.go -i
Goby 0.0.9
» String.is_a
ArgumentError: Expect 1 argument. got=0
#=> ERROR: ArgumentError: Expect 1 argument. got=0
» String.is_a(Class)
Internal Error: runtime error: index out of range
panic: runtime error: index out of range [recovered]
        panic: runtime error: index out of range

goroutine 1 [running]:
github.com/goby-lang/goby/vm.(*thread).execInstruction.func1(0xc42011b860)
        /Users/MaxwellAlexius/go/src/github.com/goby-lang/goby/vm/thread.go:71 +0x81
panic(0x1336140, 0x15392a0)
        /usr/local/Cellar/go/1.8.3/libexec/src/runtime/panic.go:489 +0x2cf
github.com/goby-lang/goby/vm.glob..func21(0xc42011b860, 0xc4201091f0, 0xc42011f400, 0x2, 0x2)
        /Users/MaxwellAlexius/go/src/github.com/goby-lang/goby/vm/instruction.go:343 +0x452
github.com/goby-lang/goby/vm.(*thread).execInstruction(0xc42011b860, 0xc4201091f0, 0xc420128fc0)
        /Users/MaxwellAlexius/go/src/github.com/goby-lang/goby/vm/thread.go:75 +0x96
github.com/goby-lang/goby/vm.(*thread).evalCallFrame(0xc42011b860, 0xc4201091f0)
        /Users/MaxwellAlexius/go/src/github.com/goby-lang/goby/vm/thread.go:42 +0x60
github.com/goby-lang/goby/vm.(*thread).startFromTopFrame(0xc42011b860)
        /Users/MaxwellAlexius/go/src/github.com/goby-lang/goby/vm/thread.go:36 +0x55
github.com/goby-lang/goby/vm.(*VM).startFromTopFrame(0xc420102240)
        /Users/MaxwellAlexius/go/src/github.com/goby-lang/goby/vm/vm.go:190 +0x2f
github.com/goby-lang/goby/vm.(*VM).REPLExec(0xc420102240, 0xc42011ceb0, 0x2, 0x2)
        /Users/MaxwellAlexius/go/src/github.com/goby-lang/goby/vm/repl.go:42 +0x64e
github.com/goby-lang/goby/igb.StartIgb(0x139a4df, 0x5)
        /Users/MaxwellAlexius/go/src/github.com/goby-lang/goby/igb/repl.go:212 +0xfed
main.main()
        /Users/MaxwellAlexius/go/src/github.com/goby-lang/goby/goby.go:27 +0x657
exit status 2
```

I need some time to check out for the bug

File: vm/class_test.go [hunk 2]
```diff
@@ -453,4 +462,123 @@ func TestRequireFail(t *testing.T) {
...
+	testsFail := []struct {
+		input   string
+		errType string
+		errMsg  string
+	}{
+		{`123.is_a`, ArgumentError, "ArgumentError: Expect 1 argument. got=0"},
+		{`Class.is_a`, ArgumentError, "ArgumentError: Expect 1 argument. got=0"},
+		{`123.is_a(123, 456)`, ArgumentError, "ArgumentError: Expect 1 argument. got=2"},
+		// TODO: Fix is_a Class test case
+		//{`123.is_a(Integer, String)`, ArgumentError, "ArgumentError: Expect 1 argument. got=2"},
```
Comment (Maxwell-Alexius) [inline vm/class_test.go]: The problem for implementing this test case is stated in #279

File: vm/class_test.go (hunk 1 above)
Comment (Maxwell-Alexius) [inline vm/class_test.go]: The problem for implementing this test case is stated in #279

**Comment by st0012** [comment]
@Maxwell-Alexius Please check if #280 fixes those issues so we can merge it and let you continue you work 😄

**Comment by hachi8833** [comment]
Just adding the cases of `false.is_nil #=> true` would be better.

**Comment by Maxwell-Alexius** [comment]
I'm not sure for it, because in original Ruby, `false.is_nil` is still `false`, because `nil` means empty, which is in the state neither `true` nor `false`?

**Comment by hachi8833** [comment]
Sorry, you are right. Forget it.

File: vm/class.go [hunk 3]
```diff
@@ -371,6 +371,74 @@ func builtinCommonInstanceMethods() []*BuiltInMethodObject {
...
+					c := args[0]
+					gobyClass, ok := c.(*RClass)
+
+					if !ok {
+						return t.vm.initErrorObject(TypeError, WrongArgumentTypeFormat, classClass, c.Class().Name)
+					}
+
+					receiverClass := receiver.Class()
+
+					for {
```
Comment (st0012) [inline vm/class.go]: I like the way you implement this, quite neat. Maybe you can try writing recursive anonymous function next time 👍

File: vm/class_test.go [hunk 4]
```diff
@@ -434,6 +442,7 @@ func TestRequireSuccess(t *testing.T) {
 	vm := initTestVM()
 	evaluated := vm.testEval(t, input)
 	checkExpected(t, 0, evaluated, ".rb")
+	vm.checkCFP(t, 0, 0)
```
Comment (st0012) [inline vm/class_test.go]: Thanks for adding these checking method

File: vm/class_test.go [hunk 5]
```diff
@@ -453,6 +462,101 @@ func TestRequireFail(t *testing.T) {
...
+	}
+}
+
+func TestGeneralIsNilMethodFail(t *testing.T) {
+	testsFail := []struct {
+		input  string
+		errMsg string
+	}{
+		{`123.is_nil("Hello")`, "ArgumentError: Expect 0 argument. got: 1"},
+		{`123.is_nil("Hello", "World")`, "ArgumentError: Expect 0 argument. got: 2"},
```
Comment (st0012) [inline vm/class_test.go]: I think we don't need to test so many cases for this method, especially for error conditions.

Comment (Maxwell-Alexius) [inline vm/class_test.go]: Okay, I will reduce this kind of test case

**Pull request merged**